planning/behavior_velocity_planner/autoware_behavior_velocity_blind_spot_module/src/time_to_collision.cpp

```diff
@@ -0,0 +1,239 @@
+    const double dist = autoware_utils_geometry::calc_distance2d(p1, p2);
+    const double average_velocity =
+      (p1.point.longitudinal_velocity_mps + p2.point.longitudinal_velocity_mps) / 2.0;
+    const double passing_velocity = std::max(average_velocity, 0.01);
+    passing_time += dist / passing_velocity;
+
+    future_positions.emplace_back(FuturePosition{path.points.at(i), passing_time});
+  }
+  return future_positions;
+}
+
+std::vector<FuturePosition> calculate_future_profile(
+  const autoware_internal_planning_msgs::msg::PathWithLaneId & path, const double time_to_restart,
+  const std::shared_ptr<const PlannerData> & planner_data, const lanelet::Id lane_id)
+{
+  const double nearest_dist_threshold = planner_data->ego_nearest_dist_threshold;
+  const double nearest_yaw_threshold = planner_data->ego_nearest_yaw_threshold;
+  const auto & current_pose = planner_data->current_odometry->pose;
+  const double current_velocity = planner_data->current_velocity->twist.linear.x;
+
+  const auto closest_idx = autoware::motion_utils::findFirstNearestIndexWithSoftConstraints(
+    path.points, current_pose, nearest_dist_threshold, nearest_yaw_threshold);
+
+  autoware_internal_planning_msgs::msg::PathWithLaneId reference_path;
+  bool assigned_lane_found = false;
+  for (unsigned i = 0; i < path.points.size(); ++i) {
+    auto reference_point = path.points.at(i);
+    // assume backward velocity is current ego velocity
+    if (i < closest_idx) {
+      reference_point.point.longitudinal_velocity_mps = static_cast<float>(current_velocity);
+    }
+    reference_path.points.push_back(reference_point);
+    const bool has_objective_lane_id = has_lane_ids(path.points.at(i), lane_id);
+    if (assigned_lane_found && !has_objective_lane_id) {
+      break;
+    }
+    assigned_lane_found = has_objective_lane_id;
+  }
+  if (reference_path.points.size() < 3 || !assigned_lane_found) {
+    return {};
+  }
+  auto smoothed_reference_path = reference_path;
+  if (!smoothPath(reference_path, smoothed_reference_path, planner_data)) {
+    return {};
+  }
+  return calculate_future_profile_impl(
+    smoothed_reference_path, current_pose, time_to_restart, nearest_dist_threshold,
+    nearest_yaw_threshold);
+}
+
+std::optional<std::pair<double, double>> compute_time_interval_for_passing_line(
+  const std::vector<FuturePosition> & future_positions,
+  const autoware_utils_geometry::LinearRing2d & footprint, const lanelet::ConstLineString3d & line1,
+  const lanelet::ConstLineString3d & line2)
+{
+  // search forward
+  std::optional<double> entry_time{};
+  for (const auto & [path_point, time] : future_positions) {
+    const auto & base_pose = path_point.point.pose;
+    const auto path_point_footprint =
+      autoware_utils::transform_vector(footprint, autoware_utils::pose2transform(base_pose));
+    if (boost::geometry::intersects(
+          path_point_footprint, lanelet::utils::to2D(line1).basicLineString())) {
```

Comment:
(perf) from a quick profiling, it seems the 2 `boost::geometry::intersects` account for ~60% of the runtime of the module.
I think putting all the pairs `(path_point_footprint, time)` in a Rtree and querying all the intersections with the lines would be much faster.

This is low priority and should be kept for a future PR.